lib/Optimizer/Lower/Lower.cpp

```diff
@@ -159,6 +159,76 @@ static void lowerRegressionGradNode(Function *F, CompilationContext &cctx,
                        expG);
 }
 
+static void lowerGemmNode(Function *F, CompilationContext &cctx,
+                          const GemmNode &GN) {
+
+  LOG_SCOPE(F->getLogContext(), "lowerGemm")
+
+  NodeValue A = GN.getA();
+  NodeValue B = GN.getB();
+  NodeValue C = GN.getC();
+  NodeValue Y = GN.getResult();
+  float alpha = GN.getAlpha();
+  float beta = GN.getBeta();
+
+  // Transpose A (if required).
+  if (GN.getTransposeA()) {
+    A = F->createTranspose(DECORATE_NODE_NAME(GN, "TransposeA"), A, {1, 0});
+  }
+
+  // Transpose B (if required).
+  if (GN.getTransposeB()) {
+    B = F->createTranspose(DECORATE_NODE_NAME(GN, "TransposeB"), B, {1, 0});
+  }
+
+  // If Gemm is same as FullyConnected then lower to a FullyConnected node.
+  if (isGemmSameAsFullyConnected(&GN)) {
+    NodeValue newY =
+        F->createFullyConnected(GN.getName().str(), A, B, C, Y.getType());
+    replaceAllUsesOfWith(cctx.loweredInfoMap, GN.getResult(), newY);
+    return;
+  }
+
+  // Create MatMul for A * B.
+  NodeValue newY =
+      F->createMatMul(DECORATE_NODE_NAME(GN, "MatMul"), Y.getType(), A, B);
+
+  // Multiply with alpha (if required).
+  if (alpha != 1.0) {
+    auto *alphaSplat = F->createSplat(DECORATE_NODE_NAME(GN, "AlphaSplat"),
+                                      Y.getType(), alpha);
+    newY = F->createMul(DECORATE_NODE_NAME(GN, "AlphaMul"), alphaSplat, newY);
+  }
+
+  // Check if C operand is used.
+  bool isUsedC = C.getNode();
+  if (isUsedC) {
+    auto *splatC = llvm::dyn_cast<SplatNode>(C.getNode());
+    if (splatC && (splatC->getValue() == 0)) {
```

Comment:
FYI if constant folding occurs prior to this you could end up with a Constant that is broadcasted. If this is a potential issue you may want to check for a broadcasted Constant similar to how `isConstBroadcasted()` is used elsewhere.

Reply:
I agree this case is useful to catch/treat. But I don't think I would want to put the logic here because the lowering logic is already pretty convoluted. Do you think I could detected if the `C` operand is broadcasted in Graph.cpp when creating the GemmNode (and if it is then I could directly create the GemmNode using the slice of the 'C' operand). This way also the branch which lowers the GemmNode directly to FullyConnected could benefit from this (which assumes the `C` operand is 1D). WDYT?